LFO

jpcima opened the pull request:
This adds SFZv2 LFO on top of the mod matrix PR.

There are the basic targets for now (amplitude, pitch, etc)

Simple usage example
```
<region>
sample=*sine
lfo1_freq=1.0
//lfo1_amplitude=100
lfo1_pitch=100
```

jpcima committed:
Add SFZv2 LFO and a small set of modulation targets

jpcima committed:
Add the LFO plotting tool

paulfd committed:
Const stuff

paulfd commented on src/sfizz/LFO.cpp:
```diff
@@ -0,0 +1,295 @@
+}
+
+template <LFOWave W>
+void LFO::processSH(unsigned nth, absl::Span<float> out)
```
These functions share alot of code, I wonder if we could avoid this repetition. I can try to work on this unless you have an idea.

jpcima replied:
Repetitions, in particular `processWave`, it's not just for avoid the branch in the frame loop, also I wish for SSE impls of wave functions. All are vectorizable.

paulfd commented on src/sfizz/LFO.cpp:
```diff
@@ -0,0 +1,295 @@
+    const float fadeInPole = impl.fadeInPole_;
+    for (size_t i = 0; i < numFrames; ++i) {
+        out[i] *= fadeIn;
+        fadeIn = fadeInPole * fadeIn + (1 - fadeInPole);
```
This is the fade in used in ARIA/Cakewalk?

jpcima replied:
Fade is too slow, I'll adjust the ratio to make this better.

Sfizz-Cakewalk and Sfizz-Aria comparison
![sfizz-cakewalk](https://user-images.githubusercontent.com/17614485/89967668-5cc40280-dc52-11ea-9e8a-55bcccd4a44b.png)
![sfizz-aria](https://user-images.githubusercontent.com/17614485/89967688-66e60100-dc52-11ea-9d0c-b65b3c3ad9c1.png)

jpcima replied:
As it turns out, the fade-in is just a linear progression. Implemented in latest.
(as a note, it would be trivial to let fade-in be defined by curve, if that might be desired)

paulfd committed:
Typo

paulfd commented on src/sfizz/LFO.cpp:
```diff
@@ -0,0 +1,295 @@
+    for (size_t i = 0; i < numFrames; ++i) {
+        out[i] += offset + scale * sampleHoldValue;
+
+        // TODO(jpc) lfoN_count: number of repetitions
```
Is there more than counting the cumulative number of numWraps?

jpcima replied:
It's what it's supposed to be. In Aria, this is unimplemented, and I never tried it in Cakewalk.
The reason of omitting it for time being, it's vectorizability of this whole loop is challenged by this conditional branch.
(but surely there are some workable alternative solutions)

paulfd committed:
Const stuff

paulfd commented on src/sfizz/Region.cpp:
```diff
@@ -1324,3 +1546,22 @@ float sfz::Region::getBendInCents(float bend) const noexcept
+    c.target = target;
+
+    connections.push_back(c);
+    return connections.back();
```
Why don't we have this as a return value of push_back/emplace_back in C++11, I'll never understand...

jpcima replied:
same

paulfd commented on src/sfizz/Voice.cpp:
```diff
@@ -34,6 +35,10 @@ sfz::Voice::Voice(int voiceNumber, sfz::Resources& resources)
         filter.setGain(vaGain(config::filteredEnvelopeCutoff, sampleRate));
 }
 
+sfz::Voice::~Voice()
```
Any reason to have this?

jpcima replied:
This was a recommendation of google c++ guidelines.
This would ensure destructions don't get inlined in the caller, for economy of code size.

paulfd commented on src/sfizz/Region.cpp:
```diff
@@ -786,6 +786,228 @@ bool sfz::Region::parseOpcode(const Opcode& rawOpcode)
+            if (!extendIfNecessary(lfos, lfoNumber, Default::numLFOs))
+                return false;
+            if (auto value = readOpcode(opcode.value, Default::lfoPhaseRange)) {
+                float normalPhase = *value * (1.0 / 360.0);
```
Are you sure about this? The sfzformat page seems to contradict this.

jpcima replied:
Phase is definitely expressed in degree unit.

phase=0
![LFO phase=0](https://user-images.githubusercontent.com/17614485/89971676-36a36000-dc5c-11ea-91a9-4f80a3958708.png)

phase=90
![LFO phase=90](https://user-images.githubusercontent.com/17614485/89971685-3b681400-dc5c-11ea-8a13-e258a24daa45.png)

paulfd committed:
Initialize the LFOs in sfizz_plot_lfo

paulfd committed:
Add LFO regression tests
The reference files have been visually validated

Tweaks for the LFO regression test

paulfd commented:
I added some tests, trying to automate them also on travis. It may fail, bear with me :)

jpcima committed:
Implement the correct LFO fade in

jpcima commented:
On reference data points, may we possibly flac them up to reduce their size?

paulfd committed:
Explicitely compute the step out of the loop

paulfd reviewed (approved):
So if you're certain about the phase I guess we need to amend sfzformat.com, and if my solution for the reference sampling point is OK, I'm good with the rest :slightly_smiling_face:

jpcima commented:
Last commit introduces a terrible hack for Garritan instrument compatibility.
- if `lfoN_phase` is 0<=x<=1, it's considered with the ARIA meaning which is 0-1 normalized
- otherwise, it's taken as degrees, as standard SFZ suggests
- if the user adds trailing text to the value, this forces a particular choice of unit, eg. `0.5d` → 0.5 degrees

paulfd commented:
This seems like a reasonable solution to a messy problem!

AlexMitchellMus commented:
`lfoN_phase_onccX` would then follow the SFZv2 standard, or use both as well?

jpcima committed:
Command-line arguments for sfizz_plot_lfo

jpcima committed:
Allow the LFO tool to set options and output FLAC

jpcima commented:
I have done some changes as follows to help make the work more mergeable
- control all of `plot_lfo` settings by arguments
- squashed the test commit and moved on top
- postponed the ARIA compatibility work, kept in a side branch; there is discussion to have on this topic, as every proposed solution has its advantages and inconvenients, and I'm not interested to take the decision here, just advance LFO for coming work

On the tests, I wish to have them preferably included in the regular test suite, I will add some work towards this direction.

jpcima commented:
Regression tests are moved into the main test suite. Good for merge.
Do we keep these numpy files?

paulfd commented:
I like the plotting one but it can go somewhere else. I'm not used to whipping out gnuplots on the fly :)

We still need something like it to validate changes visually and possibly update the reference.
